_adr/88/index.adoc

```diff
@@ -1,48 +1,71 @@
-// * No undefined jargon
-
-// No need for a title heading, it's added by the template
-
 ## Context and Problem Statement
-// What is the background against which this decision is being taken?
+The purpose of this ADR is to define a standardized approach for securing admin API in a fleet manager
 
 ## Goals
-// Bulleted list of outcomes that this ADR, if accepted, should help achieve
+* Provide standardized way to secure admin API endpoints for fleet managers
 
 ## Non-goals
-// Bulleted list of outcomes that this ADR is not trying to achieve.
 
 ## Current situation
-// Where are we now?
 
 ## Proposal
-// What is the decision being proposed
+
+image::adr-88-admin-api-architecture-diagram.png[architecture diagram]
+
+A Keycloak realm is used to authorize access to admin API endpoint(s). 
+Such realm contains roles that allow access to endpoints to the privileged API users.
+Recommended approach would be to create a set of realm roles per environment and name them in the way that indicates the permissions they grant as follows:
+
+* <unique-fleet-manager-prefix>-admin-read-stage
+* <unique-fleet-manager-prefix>-admin-write-stage
+* <unique-fleet-manager-prefix>-admin-full-stage
+
+* <unique-fleet-manager-prefix>-admin-read-prod
+* <unique-fleet-manager-prefix>-admin-write-prod
+* <unique-fleet-manager-prefix>-admin-full-prod
+
+These roles map to the following permissions for the environment included as the role postfix (e.g. `prod`):
+
+* \*-read-* - list permissions
+* \*-write-* - list/ update permissions
+* \*-full-* - full permissions (list/ update/ delete)
+
+Example full role for kas-fleet-manager in production environment is: `kas-fleet-manager-admin-full-prod`.
+
+Once Keycloak instance is set up, a privileged user uses their credentials to get a JWT token from Keycloak.
+The JWT token is sent to a admin API in a HTTP request.
+Admin API authorization middleware verifies token signature with Public JWKS Certificate and roles included in the JWT token are checked against roles specified in the fleet manager.
+Matching role for given admin endpoint grants access to the action associated with that endpoint.
+Failure to match the required role results in restricting access to the endpoint.
+
+image::adr-88-authentication-flow.png[architecture diagram]
+
+## Suggested implementation for Red Hat Internal Keycloak
+
+Instead of using dedicated Keycloak instance/ realm, roles allowing to access admin API endpoints are represented by membership in a Red Hat Rover Group(s).
 
 ### Threat model
-// Provide a link to the relevant threat model. 
-// You must either update an existing threat model(s) to cover the changes made by this ADR, or add a new threat model.
+Admin API endpoints are protected by realm roles, hence only users that are assigned these roles can access the endpoints.
+
+No additional mitigation is required.
 
 ## Alternatives Considered / Rejected
 
 ## Challenges
-// What are the costs/drawbacks of the proposed decision?
```

Comment:
Surely there are some: Lots of Rover groups needed for large numbers of customers? IDK if this role membership information is imported into keycloak, or queried in-place in Rover, but if the latter then what are the consequences wrt availability?

Reply:
@tombentley What do you mean by "a large number of customers?". SRE group is the only consumer of these endpoints. Roles == rover groups, roles are created in keycloak. When a user request tokens, these roles are added as claims in tokens. Per service, we will have 6 roles (3 stage, 3 prod). If the number of services increases to 50, then roles will be 300; if the same set of users is added in all 300 groups, it increases the token size, impacting the admin endpoints workflow. Either request will get blocked at the gateway due to token size or time to parse/decode token will increase the response time.

Reply:
> SRE group is the only consumer of these endpoints.

That was the part I was missing, thanks.

But my point was really more general: Surely there are _some_ pros and cons to what's proposed?